### PR TITLE
dnsproxy: Improve regex pattern

### PR DESCRIPTION
**Pull request by odinuge**
_TL;DR:_ This is now V3 of https://github.com/cilium/cilium/pull/20214.

The full list of changes can be found in the commits (tho. this is based off https://github.com/cilium/cilium/pull/21288);
- Extract regex anchors to the beginning and end of the full pattern
- Remove the use of regex groups since we don't use them
- Short circuit matchPatterns with just a wildcard and match all fqdns
- Skip running `.MatchString` on wildcard regexes.

The overall perf change for all commits in this PR vs. https://github.com/cilium/cilium/pull/21288 (all without the lru cache, and ignore the time/op, since they are running in docker for mac 😢 ):

```
$ benchstat <pre-this-pr> <this-pr>
name                             old time/op          new time/op          delta
_perEPAllow_setPortRulesForID-7           77.4s ± 8%           83.0s ±17%     ~     (p=0.310 n=5+5)

name                             old B(HeapInUse)/op  new B(HeapInUse)/op  delta
_perEPAllow_setPortRulesForID-7           97.3M ± 1%           74.2M ± 1%  -23.75%  (p=0.008 n=5+5)

name                             old alloc/op         new alloc/op         delta
_perEPAllow_setPortRulesForID-7          78.6GB ± 0%          73.0GB ± 0%   -7.18%  (p=0.008 n=5+5)

name                             old allocs/op        new allocs/op        delta
_perEPAllow_setPortRulesForID-7            167M ± 0%            108M ± 0%  -35.37%  (p=0.008 n=5+5)
```
---

After this we plan doing;
- Switch from string concatenation to a byte buffer in `GeneratePattern`
- Consider the regex for matchPatterns + `map[string]struct{}` for matchNames.
- Replacing the `[.]` with `\.` to reduce the regex size
- Replacing the `*` replacement from `[-a-zA-Z0-9_]*` to `[^.]*`
- ...

--

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #issue-number

```release-note
dnsproxy: Improve regex used for matching dns queries by reducing its complexity and size to save memory and speed up matching
```

### CODE REVIEW

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR! It would be great if we can get benchmarking code added to the tests. There should already be a framework that tests at a high-level inside `pkg/fqdn/dnsproxy/proxy_test.go`. However, it would be good to add more piece-meal benchmarks corresponding to the code changes in this PR.

`Benchmark_perEPAllow_setPortRulesForID_large` is skipped by default, but if you remove the `t.Skip()` and add the CNPs as a yaml to `testdata/cnps-large.yaml` and run go test, you should be able to get some results.

**Comment by odinuge** [comment]
Hey @christarazi! Thanks! Did the initial testing on the v1.9 release branch, since thats what we are using atm., so didn't see those benchmarks.

I don't have a set of policies at hand atm., I've just been creating a set of domains to push through. Do you have the CNP list you were using in that PR?

Also, discussed this with a colleague, and we decided to try using a set (essentially a `map[string]bool`) for the fqdn names, and then keeping the regex for patterns. That was waaaaay faster on the synthetic tests as well, and the memory allocations aren't even comparable. Will need to test in some real environment first tho. Will hopefully do some quick implementation this week, and will ship my benchmarking code as soon as I can, hopefully before end of week.

**Comment by odinuge** [comment]
Did a quick test now with this PR vs. master, with a lru size of 128 with the same dataset as @christarazi was using with his initial testing;

```
# This PR
$ go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before (N=1)
Alloc = 32 MiB  HeapInuse = 53 MiB      Sys = 982 MiB   NumGC = 15
After (N=1)
Alloc = 369 MiB HeapInuse = 378 MiB     Sys = 982 MiB   NumGC = 37
Before (N=1)
Alloc = 34 MiB  HeapInuse = 52 MiB      Sys = 1017 MiB  NumGC = 48
After (N=1)
Alloc = 237 MiB HeapInuse = 300 MiB     Sys = 1017 MiB  NumGC = 71
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        5772238001 ns/op        2161473136 B/op 23052832 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      21.727s
```
```
# Baseline / master
root@eab730ccc555:/code# go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before (N=1)
Alloc = 32 MiB  HeapInuse = 50 MiB      Sys = 990 MiB   NumGC = 14
After (N=1)
Alloc = 464 MiB HeapInuse = 484 MiB     Sys = 990 MiB   NumGC = 35
Before (N=1)
Alloc = 36 MiB  HeapInuse = 54 MiB      Sys = 1101 MiB  NumGC = 45
After (N=1)
Alloc = 527 MiB HeapInuse = 538 MiB     Sys = 1101 MiB  NumGC = 66
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        5331359995 ns/op        3019854640 B/op 26485069 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      23.638s
```

We clearly see the heap usage is very much lower with this PR. Still need some tweaking tho. Will try to grab a bigger dataset and check the diff

File: pkg/fqdn/dnsproxy/proxy.go [hunk 1]
```diff
@@ -176,7 +177,7 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
 	}
 
 	for i := range ipRules {
-		if _, exists := ipRules[i].IPs[destIP]; (exists || ipRules[i].IPs == nil) && ipRules[i].Re.MatchString(name) {
+		if _, exists := ipRules[i].IPs[destIP]; (exists || ipRules[i].IPs == nil) && ipRules[i].Re.MatchString(Reverse(name)) {
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: We probably would want to reverse the `name` once here ideally rather than on each `ipRules`. We can pass in a reversed `name` to `checkRestored` or do the reverse before the for-loop.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 2]
```diff
@@ -552,7 +553,7 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 
 	for selector, re := range epAllow {
 		// The port was matched in getPortRulesForID, above.
-		if selector.Selects(destID) && re.MatchString(name) {
+		if selector.Selects(destID) && re.MatchString(Reverse(name)) {
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: Same as above re: reversing once rather multiple times needlessly

File: pkg/fqdn/dnsproxy/proxy.go [hunk 3]
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
 		reStrings := make([]string, 0, len(l7Rules.DNS))
 		for _, dnsRule := range l7Rules.DNS {
 			if len(dnsRule.MatchName) > 0 {
-				dnsRuleName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
-				dnsPatternAsRE := matchpattern.ToRegexp(dnsRuleName)
-				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+				dnsRuleName := FromFqdn(dnsRule.MatchName)
+				dnsPatternAsRE, _ := CustomToRegexp(Reverse(dnsRuleName))
+				reStrings = append(reStrings, dnsPatternAsRE)
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: Do the `matchName`s need to be reversed? From what I understand, reversing the string makes sense if it's a pattern which contains multi-char matches such as `*` or `.+` type matches, i.e. `*.*.svc.cluster.local` because that allows the regex engine to "exit early". To elaborate, every DNS name comparison against the regex that starts with `*` and `*` will force the engine to go through all possible matches upfront, rather than when necessary.

For example, take the pattern `*.*.svc.cluster.local` and a DNS name `foo.bar.svc.cluster.local`. Without reversing, the regex engine will first process `foo` and `bar` and they'll match against the corresponding `*`'s. However, if the DNS name instead was `foo.bar.something.else.local`, then only after the `*` checks, does the regex engine realize this DNS name doesn't match.

Let's say we take the reverse of the pattern `local.cluster.svc.*.*` and `foo.bar.something.else.local`. The first `local` part will pass as a basic string match (less expensive than wildcard presumably). The second part, `else`, matches against `cluster` and the engine returns false, whereas if it was `*` it would've return true at this part of the domain.

If it's a name, then I don't see any improvement related to the above which can be possible, so we are probably just wasting cycles reversing the string.

(I think it would be helpful to put this as a code comment if accurate to explain why we do this.)

File: pkg/fqdn/dnsproxy/proxy.go [hunk 4]
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
...
-				dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
-				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+				dnsPattern := FromFqdn(dnsRule.MatchPattern)
+				dnsPatternAsRE, isAllAllowed := CustomToRegexp(Reverse(dnsPattern))
+				// If the pattern is a wildcard, quit the regex calculation and insert a simple allow
+				// all rule. This will reduce memory usage and improve performance significantly when
+				// a wildcard rule is present together with a set of other rules.
+				// TODO do even need any validation at all here, since we will only ever get valid domains?
+				// eg do only "."
+				if isAllAllowed {
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]:
```suggestion
				dnsPatternAsRE, isWildcard := CustomToRegexp(Reverse(dnsPattern))
				// If the pattern is a wildcard, quit the regex calculation and insert a simple allow
				// all rule. This will reduce memory usage and improve performance significantly when
				// a wildcard rule is present together with a set of other rules.
				// TODO do even need any validation at all here, since we will only ever get valid domains?
				// eg do only "."
				if isWildcard {
```

Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy.go]: Can you elaborate on the TODO?

> TODO do even need any validation at all here, since we will only ever get valid domains?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 5]
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
...
+				reStrings = append(reStrings, dnsPatternAsRE)
 			}
 		}
-		mp := strings.Join(reStrings, "|")
-		rei, err := re.CompileRegex(mp)
+		// Sort the rules to make faster and smaller regexes, as well as speeding up the compilation.
+		// The impact of sorting is negligible for a low number of rules, and the more rules, the more the
+		// benefit of lower memory usage, faster compilation and faster matching start being noticeable
+		// TODO test this in real life to see for performance improvements
+		sort.Strings(reStrings)
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]:
```suggestion
		if !isWildcard {
		  sort.Strings(reStrings)
		}
```

Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy.go]: Actually, we might want to separate out this change from the rest in a different commit. That way we can get benchmark results with this PR and then with this PR + sorting, so we can easily tell what sort of improvement / regression sorting might have.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 6]
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
...
+		// Explicitly add end and beginning of string markers to the beginning and end to avoid having them
+		// defined on a per dns rule basis. Also avoid wrapping each dns rule inside a group since we don't
+		// need to know where a given match occur. All those will also improve the performance of both the
+		// regex compilation and the regex matching
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: Makes sense as something to be part of the commit msg.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 7]
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
...
+		// TODO figure out if we can/want use a separate key for the regex cache to avoid the sorting
+		// and extra logic in case its present in the cache
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: What do you have in mind as keys to the cache?

File: pkg/fqdn/dnsproxy/proxy.go [hunk 8]
```diff
@@ -930,3 +956,47 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
 
 	return newRE, nil
 }
+
+// Removes the last trailing dot in case its an fqdn
+// TODO move to the proper location
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: These can probably be moved to the `pkg/fqdn/matchpattern` pkg as there exists similar code to this there.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 9]
```diff
@@ -930,3 +956,47 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
...
+func Reverse(s string) string {
+	size := len(s)
+	buf := make([]byte, size)
+	for start := 0; start < size; {
+		r, n := utf8.DecodeRuneInString(s[start:])
+		start += n
+		utf8.EncodeRune(buf[size-start:], r)
+	}
+	return string(buf)
+}
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: Would be good to write a benchmark test for this function so that we can evaluate other implementations and compare them.

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 10]
```diff
@@ -800,17 +800,19 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
...
 			"IPs": {"::": {}}
 		}, {
-			"Re":  "(^cilium[.]io[.]$)",
+			"Re":  "^[.](?:oi[.]muilic)$",
 			"IPs": {"127.0.0.1": {}, "127.0.0.2": {}}
 		}],
 		"54": [{
-			"Re":  "(^example[.]com[.]$)",
+			"Re":  "^[.](?:moc[.]elpmaxe)$",
 			"IPs": null
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy_test.go]: One thing we can do if we are worried about the readability of the tests is to construct this string with `fmt.Sprintf`. I don't think this is too big of a deal, specifically as we have a comment explaining why it looks unexpected.

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR, this is a great step towards a massive performance improvement in the Agent!

I understand this is a WIP :), and I have several comments about the code, but overall the idea is sound!

**Comment by aanm** [comment]
@odinuge I've marked the PR as draft since it's in WIP. Feel free to mark it as ready for review once it's done. Thank you!

File: pkg/fqdn/dnsproxy/proxy.go (hunk 3 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Hey!

> [...]
> If it's a name, then I don't see any improvement related to the above which can be possible, so we are probably just wasting cycles reversing the string.

Yee, very much agree. Ref. https://github.com/cilium/cilium/pull/20396 I think we can keep it as is for now, since that PR will likely be merged first! 😄 Also, if we wanted to reverse matchPatterns and not matchNames, they would either have to be in separate regexes, or we would have had to run `Match` on the domain twice; one each way.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 4 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Yeee. My thought is that in case we hit a wildcard, we don't really care about the regex to begin with. in that case, can we just return `^.*$` or something to always short circuit the regex?

And thanks for the suggestion, that looks better!

File: pkg/fqdn/dnsproxy/proxy.go (hunk 6 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Yee, sounds good! Will do the final touches of the commit msg when we converge towards the desired state. 👌

File: pkg/fqdn/dnsproxy/proxy.go (hunk 8 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Sounds good, thanks!

File: pkg/fqdn/dnsproxy/proxy.go (hunk 7 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Lets discuss all this in https://github.com/cilium/cilium/pull/20396 😄

File: pkg/fqdn/dns/dns.go [hunk 11]
```diff
@@ -37,3 +37,13 @@ func FQDN(s string) string {
 	}
 	return strings.ToLower(s) + "."
 }
+
+// FromFQDN returns the fqdn without the trailing dot.
+// If s isn't fully qualified, it behaves as the identity function.
+func FromFQDN(s string) string {
```
Comment (odinuge) [inline pkg/fqdn/dns/dns.go]: Hmmm. What should we name this?? 🤔

File: pkg/fqdn/dnsproxy/proxy.go (hunk 5 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Yee, agree!

wow, this makes much more of a diff than expected @christarazi! good call!

First one without sorting, number two with sorting; look at the heap diff
```
root@43144d6cc6bb:/code# go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 19 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 980 MiB   NumGC = 18
After Test (N=1,EPs=20,cache=128)
Alloc = 752 MiB HeapInuse = 892 MiB     Sys = 1405 MiB  NumGC = 47
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        10003844804 ns/op       5845248776 B/op 37991006 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      16.137s
root@43144d6cc6bb:/code# go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 38 MiB  HeapInuse = 49 MiB      Sys = 934 MiB   NumGC = 18
After Test (N=1,EPs=20,cache=128)
Alloc = 446 MiB HeapInuse = 586 MiB     Sys = 935 MiB   NumGC = 49
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        10052384684 ns/op       4664795352 B/op 41661104 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      14.822s
```

File: pkg/fqdn/matchpattern/matchpattern_test.go [hunk 12]
```diff
@@ -25,14 +25,28 @@ var _ = Suite(&MatchPatternTestSuite{})
...
 	for source, target := range map[string]string{
 		"cilium.io.":   "^cilium[.]io[.]$",
 		"*.cilium.io.": "^" + allowedDNSCharsREGroup + "*[.]cilium[.]io[.]$",
 		"*":            "(^(" + allowedDNSCharsREGroup + "+[.])+$)|(^[.]$)",
 		".":            "^[.]$",
 	} {
-		reStr := ToRegexp(source)
+		reStr := ToAnchoredRegexp(source)
+		_, err := regexp.Compile(reStr)
+		c.Assert(err, IsNil, Commentf("Regexp generated from pattern %sis not valid", source))
```
Comment (schlosna) [inline pkg/fqdn/matchpattern/matchpattern_test.go]:
```suggestion
		c.Assert(err, IsNil, Commentf("Regexp generated from pattern %s is not valid", source))
```

**Comment by christarazi** [comment]
Is this ready for review @odinuge?

File: pkg/fqdn/matchpattern/matchpattern.go [hunk 13]
```diff
@@ -76,3 +85,28 @@ func ToRegexp(pattern string) string {
...
+	pattern = strings.ToLower(pattern)
+	// handle the * match-all case. This will filter down to the end.
+	if pattern == "*" {
+		return MatchAllUnAnchoredPattern
+	}
+	pattern = escapeRegexpCharacters(pattern)
+	return pattern
+}
+
+func escapeRegexpCharacters(pattern string) string {
```
Comment (schlosna) [inline pkg/fqdn/matchpattern/matchpattern.go:99]: Should `ToAnchoredRegexp` use this function as well?

https://github.com/cilium/cilium/blob/d0d9b6c9cec745192704b8518b98d67d541f4a2e/pkg/fqdn/matchpattern/matchpattern.go#L78-L83

File: pkg/fqdn/matchpattern/matchpattern_test.go [hunk 14]
```diff
@@ -25,14 +25,28 @@ var _ = Suite(&MatchPatternTestSuite{})
 // cilium.io. -> cilium[.]io[.]
 // *.cilium.io. -> [-a-zA-Z0-9]+.cilium[.]io[.]
 // *cilium.io. -> "([a-zA-Z0-9]+[.])?cilium[.]io[.]
-func (ts *MatchPatternTestSuite) TestMatchPatternREConversion(c *C) {
+func (ts *MatchPatternTestSuite) TestAnchoredMatchPatternREConversion(c *C) {
 	for source, target := range map[string]string{
 		"cilium.io.":   "^cilium[.]io[.]$",
 		"*.cilium.io.": "^" + allowedDNSCharsREGroup + "*[.]cilium[.]io[.]$",
 		"*":            "(^(" + allowedDNSCharsREGroup + "+[.])+$)|(^[.]$)",
```
Comment (schlosna) [inline pkg/fqdn/matchpattern/matchpattern_test.go:32]: Should this use non-capturing groups?

```suggestion
		"*":            "^(?:(?:" + allowedDNSCharsREGroup + `+\.)+|\.)$`,
```

File: pkg/fqdn/dnsproxy/proxy.go [hunk 15]
```diff
@@ -1123,23 +1122,26 @@ func shouldCompressResponse(request, response *dns.Msg) bool {
...
+			reStrings = append(reStrings, dnsRuleNameAsRE)
 		}
 		if len(dnsRule.MatchPattern) > 0 {
 			dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
-			dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
-			reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+			dnsPatternAsRE := matchpattern.ToUnAnchoredRegexp(dnsPattern)
+			if dnsPatternAsRE == matchpattern.MatchAllUnAnchoredPattern {
+				return matchpattern.MatchAllAnchoredPattern
+			}
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go:1124]: In an offline sync, you mentioned this was awkward. Could you explain why in a code comment?

File: pkg/fqdn/matchpattern/matchpattern_test.go (hunk 12 above)
Comment (odinuge) [reply pkg/fqdn/matchpattern/matchpattern_test.go]: ooo, good catch, thanks!

File: pkg/fqdn/matchpattern/matchpattern_test.go (hunk 14 above)
Comment (odinuge) [reply pkg/fqdn/matchpattern/matchpattern_test.go:32]: Yeah, I agree. I think its outside the scope of this PR tho., and imo we should just make that regex match everything, ala. `^.*$` or something, since by design it means _match everything. And if so, we can skip the regex `MatchString` call as well.

File: pkg/fqdn/dnsproxy/proxy.go (hunk 15 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go:1124]: 👋 

Yeah, this line effectively makes the `MatchAllUnAnchoredPattern` var only used here. Was thinking about making the `ToUnAnchoredRegexp` func return `string, bool` and `true` when its a wildcard, but this is just as verbose and easy reason about. I don't feel too much for one over the other.

Any thoughts?

File: pkg/fqdn/matchpattern/matchpattern.go (hunk 13 above)
Comment (odinuge) [reply pkg/fqdn/matchpattern/matchpattern.go:99]: Yeah, that was the plan! Thanks, updated the PR now!

File: pkg/fqdn/dnsproxy/proxy.go (hunk 15 above)
Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy.go:1124]: Ah, seems straightforward enough as-is.

**Review by christarazi — APPROVED** [review]
One thing I just realized is that once this lands, will it change the behavior of existing policies? Is there a potential it might break on existing matchpatterns in user environments?

**Comment by odinuge** [comment]
Hi, thanks!

> One thing I just realized is that once this lands, will it change the behavior of existing policies? Is there a potential it might break on existing matchpatterns in user environments?

This should be a noop for users, since what the regex will match is the same, its just the regex pattern that is different. It should also be fully backwards compatible with older versions with the restore feature as well (unless I'm missing anything obvious..).

I think the unit-tests [here](https://github.com/cilium/cilium/blob/master/pkg/fqdn/dnsproxy/proxy_test.go) and [here](https://github.com/cilium/cilium/blob/master/pkg/fqdn/dnsproxy/proxy_test.go) should cover all cases, and make sure that the end user behavior is the same.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
Travis failed with https://github.com/cilium/cilium/issues/23314

**Comment by odinuge** [comment]
Thanks @aanm!

We deployed this to production here over the last few weeks, and on the hosts with the biggest l7 rulests and the most dns queries, including a `matchPattern: *` rule for some stuff, we saw a memory reduction of more than 50% (the graph is per host/node, and the drop is when deploying a version with this patchset).


Screen Shot 2023-01-30 at 11 52 51 AM
